docs/design-documents/20220408-loading-pipeline-configuration-files.md

```diff
@@ -0,0 +1,270 @@
+# Loading Pipeline Configuration Files
+
+## Summary 
+
+This document describes the design and implementation for loading pipeline 
+configuration files. It details how Conduit loads a configuration file at start.
+
+In order to provide a detailed history of configuration changes, better secret
+handling, and increased confidence in a given configuration, we propose
+adding support for loading and subsequently running pipelines entirely from 
+configuration files loaded into Conduit at its root or a configured directory.
+
+Our solution proposes an immutable Conduit runtime backed by an in-memory DB
+implementation that merges a user-provided configuration file with environment
+variables and CLI flags at runtime to seed Conduit's services with the necessary 
+information for full operation. 
+
+## Goals 
+
+- Fully load and configure pipelines from a file.
+- Properly handle environment variable injection into that file.
+
+### Non Goals
+
+Defining what we don't want out of this design and implementation is equally 
+important as what we _do_ want.
+
+- We don't care about the structure or schema of configuration files.
+- We shouldn't implement key management or encryption for secrets.
+
+## Scope
+
+We are strictly concerned with the loading of pipeline files at start time and 
+how those files are handled through Conduit's full operation cycle.
+
+Configuration file schema and other file-level design choices are out of scope.
+
+This document doesn't make any definitions or statements about how the UI should
+handle these changes. Those should be reserved for later design documents and
+discussion.
+
+## Key Terms 
+
+**pipeline** - a Conduit pipeline with `n` number of Source connectors, 
+`m` number of Processors, and `o` number of Destination connectors.
+
+**production mode** - a mode of operation where Conduit pipelines must be
+statically defined and cannot change after Conduit starts. No UI is served in
+production mode, and only pipeline lifecycle methods can be called over the
+API.
```

Comment:
I feel like such a mode would impose a tough limitation. I would rather we limit the immutability to the level of a pipeline, essentially saying that if a certain pipeline is provisioned using a config file it can't be mutated through the API.

Reply:
I built a good chunk of this design with conversations in mind where we had discussed a stricter production-mode standard for Conduit, so this limitation is admittedly part of my design. I would be open to considering this change, though. I think it achieves nearly similar results, with perhaps broader use cases. Let me think more about it! 🤔

Reply:
I feel like this was not addressed.